core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java

```diff
@@ -0,0 +1,298 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.journal.raft;
+
+import alluxio.ProcessUtils;
+import alluxio.master.journal.AbstractCatchupThread;
+import alluxio.master.journal.CatchupFuture;
+import alluxio.master.journal.JournalEntryAssociation;
+import alluxio.master.journal.JournalUtils;
+import alluxio.master.journal.Journaled;
+import alluxio.master.journal.sink.JournalSink;
+import alluxio.proto.journal.Journal;
+import alluxio.resource.LockResource;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+/**
+ * Used to control applying to masters.
+ * It supports suspending and resuming.
+ *
+ * TODO(ggezer): Extend with on-disk buffering.
+ */
+@ThreadSafe
+public class BufferedJournalApplier {
+  private static final Logger LOG = LoggerFactory.getLogger(BufferedJournalApplier.class);
+
+  /** Resume will acquire exclusive lock when buffer has less than this value. */
+  private static final int RESUME_LOCK_BUFFER_SIZE_WATERMARK = 100;
+
+  /** Resume will acquire exclusive lock when can't finish after this much time. */
+  private static final int RESUME_LOCK_TIME_LIMIT_MS = 30000;
+
+  /** Journals managed by this applier. */
+  private final Map<String, RaftJournal> mJournals;
+  /** A supplier of journal sinks for this applier. */
+  private final Supplier<Set<JournalSink>> mJournalSinks;
+
+  /** The last sequence applied. */
+  private long mLastAppliedSequence = -1;
```

Comment:
Do we need to track one sequence number for each journal?

Reply:
This is used by `JournalStateMachine` which manages all journals with a single sequence. So not required at here as well.